akka-persistence/src/main/scala/akka/persistence/journal/JournalProxy.scala

```diff
@@ -0,0 +1,181 @@
+      context.setReceiveTimeout(1.second) // for retries
+      context.become(identifying(address))
+    case InitTimeout ⇒
+      log.info("Initialization timeout, Use `JournalProxy.setTargetLocation`")
+      context.become(initTimedOut)
+      unstashAll() // will trigger appropriate failures
+    case msg ⇒
+      stash()
+  }
+
+  def becomeIdentifying(address: Address): Unit = {
+    sendIdentify(address)
+    context.setReceiveTimeout(1.second) // for retries
+    context.become(identifying(address))
+  }
+
+  def sendIdentify(address: Address): Unit = {
+    val sel = context.actorSelection(RootActorPath(address) / "system" / targetPluginId)
+    log.info("Trying to identify target {} at {}", pluginType.qualifier, sel)
+    sel ! Identify(targetPluginId)
+  }
+
+  def identifying(address: Address): Receive = ({
+    case ActorIdentity(`targetPluginId`, Some(target)) ⇒
+      log.info("Found target {} at [{}]", pluginType.qualifier, address)
+      context.setReceiveTimeout(Duration.Undefined)
+      unstashAll()
+      context.become(active(target, address == selfAddress))
+    case _: ActorIdentity ⇒ // will retry after ReceiveTimeout
+    case ReceiveTimeout ⇒
+      sendIdentify(address)
+  }: Receive).orElse(init)
+
+  def active(targetJournal: ActorRef, targetAtThisNode: Boolean): Receive = {
+    case TargetLocation(address) ⇒
+      if (targetAtThisNode && address != selfAddress)
+        becomeIdentifying(address)
+    case InitTimeout ⇒
+    case msg ⇒
+      targetJournal.forward(msg)
+  }
+
+  def initTimedOut: Receive = {
+
+    case req: JournalProtocol.Request ⇒ req match { // exhaustive match
+      case WriteMessages(messages, persistentActor, actorInstanceId) ⇒
+        persistentActor ! WriteMessagesFailed(timeoutException)
+        messages.foreach {
+          case a: AtomicWrite ⇒
+            a.payload.foreach { p ⇒
+              persistentActor ! WriteMessageFailure(p, timeoutException, actorInstanceId)
+            }
+          case r: NonPersistentRepr ⇒
+            persistentActor ! LoopMessageSuccess(r.payload, actorInstanceId)
+        }
+      case ReplayMessages(fromSequenceNr, toSequenceNr, max, persistenceId, persistentActor) ⇒
+        persistentActor ! ReplayMessagesFailure(timeoutException)
+      case DeleteMessagesTo(persistenceId, toSequenceNr, persistentActor) ⇒
+        persistentActor ! DeleteMessagesFailure(timeoutException, toSequenceNr)
+    }
+
+    case req: SnapshotProtocol.Request ⇒ req match { // exhaustive match
+      case LoadSnapshot(persistenceId, criteria, toSequenceNr) ⇒
+        sender() ! LoadSnapshotResult(None, toSequenceNr)
+      case SaveSnapshot(metadata, snapshot) ⇒
+        sender() ! SaveSnapshotFailure(metadata, timeoutException)
+      case DeleteSnapshot(metadata) ⇒
+        sender() ! DeleteSnapshotFailure(metadata, timeoutException)
+      case DeleteSnapshots(persistenceId, criteria) ⇒
+        sender() ! DeleteSnapshotsFailure(criteria, timeoutException)
+    }
```

Comment:
nice; takes care of the "but my PersistentActor is hanging and does nothing" nicely.